### PR TITLE
Fix minimum difficulty bugs

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

There are a bunch of bugs in Zebra's difficulty implementation, and the Zcash difficulty specifications also have some bugs.

## Solution

New changes:
- [x] Revise the testnet minimum difficulty code based on `zcashd`'s implementation zcash/zips#416
  - ZIPs 205 and 208 are wrong 
  - the min-difficulty rules change difficulty adjustment, not the difficulty filter
  - all blocks pass the default difficulty filter
- [x] Turn `expect("minimum difficulty activation is checked earlier in the test")` into an `Err` if the rule is not active (and so the function returns `None`)

Tests for spec errors:
  - [x] Test that testnet min-difficulty rules follow `zcashd` https://github.com/zcash/zips/issues/416
  - [x] Test implicit conversion of the minimum difficulty threshold to the compact representation #1277
    - covered by the genesis difficulty equality test

Reviewed changes:
- Refactor block target spacing into NetworkUpgrade methods
- Make ExpandedDifficulty Debug byte order match Hash
  - Bug fix on PR #1171
  - Without this change, test failures are really confusing
- Round-trip the PoWLimit through the compact representation
  - the Zcash specification is ambiguous, see #1277
- Add minimum difficulty testnet block vectors
- Add tests for testnet minimum difficulty blocks
  - the Zcash specification is ambiguous, see #1276
  - find a pre-Blossom and post-Blossom minimum difficulty block, and add them to `zebra_test::vectors::block`, along with their previous blocks
  - make sure the `BlockVerifier` difficulty tests pass on these new blocks

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Unit Tests and Property Tests

## Review

@yaahc has reviewed the Difficulty Contextual Validation RFC, which describes these checks.

## Related Issues

Difficulty Contextual Validation RFC PR #1246 
Difficulty tracking issue #802

Fixes bugs in:
- Previous difficulty check PR #809 
- Reverse the displayed endianness of hashes #1171
- Off-by-one error in testnet minimum difficulty block gap specification #1276 (Zcash specification)
- Implicit conversion of the minimum difficulty threshold to the compact representation #1277 (Zcash specification)
- Testnet min-difficulty rules are incorrectly specified https://github.com/zcash/zips/issues/416

## Follow Up Work

- ~Move the difficulty filter check into the state service~ - obsoleted by zcash/zips#416
- allow minimum difficulty testnet blocks
- Change the difficulty threshold check into a method on `Block` or `ExpandedDifficulty`
  - it's easy to get `<=` confused with `<`

This work is listed in the difficulty tracking issue #802

### CODE REVIEW

**Comment by teor2345** [comment]
Blocked by the `Duplicate`/`AlreadyVerified` bug #1259.

**Comment by teor2345** [comment]
Unblocked by Daira, who said that blocks 299188, 299189, and 299190 are minimum-difficulty blocks.
That's enough info to move forward with this PR, and get it out of a draft state.

I also un-assigned yaahc as the reviewer, because this PR is not ready for review yet. I'm not sure if it actually fixes the bug.

File: zebra-chain/src/parameters/network_upgrade.rs
Comment (teor2345) [inline zebra-chain/src/parameters/network_upgrade.rs]:
```suggestion
            (Network::Testnet, height) if height < TESTNET_MINIMUM_DIFFICULTY_START_HEIGHT => None,
```

File: zebra-chain/src/work/difficulty.rs
Comment (daira) [inline zebra-chain/src/work/difficulty.rs:287]: The minimum difficulty rule doesn't affect the value of `nBits`:
> As before, the `nBits` field of a minimum-difficulty block is still computed according to the original difficulty adjustment algorithm, and only this field is used for the purpose of computing the MeanTarget values from which subsequent difficulty changes are calculated.

ZIP 205 says that the target threshold is set to the value of PoWLimit for testnet. So I think this is well-specified; let me know if you disagree.

Comment (daira) [reply zebra-chain/src/work/difficulty.rs:287]: Oh, I see: https://github.com/ZcashFoundation/zebra/issues/1277

Hmm, that's technically a bug in zcashd :-/

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs]:
```suggestion
    // the minimum difficulty threshold - all blocks pass this threshold, even if
    // they aren't minimum difficulty blocks
    /// SPANDOC: Check that the testnet minimum difficulty block hash is less than the PoWLimit {?height, ?threshold, ?hash}
    assert!(hash <= ExpandedDifficulty::target_difficulty_limit(Network::Testnet),
            "testnet minimum difficulty hashes must be less than the PoWLimit");
```

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs:420]:
```suggestion
    /// SPANDOC: Check the testnet minimum difficulty start height {?height, ?hash}
    if height < block::Height(299_188) {
```

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs]:
```suggestion
    /// SPANDOC: Check that all testnet minimum difficulty block test vectors are in the list of minimum difficulty blocks {?height, ?threshold, ?hash}
```

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs]:
```suggestion
/// If `is_in_minimum_difficulty_list` is true, check the minimum difficulty rule height,
/// and make surethe previous block is in the test vectors. Otherwise, skip blocks
/// that don't pass these basic checks.
#[spandoc::spandoc]
fn check_testnet_minimum_difficulty_block(
    height: block::Height,
    is_in_minimum_difficulty_list: bool,
```

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs]:
```suggestion
    if !is_in_minimum_difficulty_list {
```

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs:424]: Delete this check after the refactor

File: zebra-chain/src/work/difficulty/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty/tests/vectors.rs]: Turn this expect into an err

**Comment by teor2345** [comment]
@yaahc after you approve this PR, I'd like to squash and rebase it myself, so we can keep some of the changes to different modules in different commits.

**Review by yaahc — DISMISSED** [review]

**Review by yaahc — DISMISSED** [review]

**Review by yaahc — APPROVED** [review]

**Pull request merged**
